src/Integration/Helpers/ToolWindowService.cs

```diff
@@ -21,45 +21,59 @@
 using System;
 using System.ComponentModel.Composition;
 using System.Diagnostics;
-using Microsoft.VisualStudio.Shell;
 using Microsoft.VisualStudio.Shell.Interop;
 using SonarLint.VisualStudio.Core;
+using SonarLint.VisualStudio.Infrastructure.VS;
 using ErrorHandler = Microsoft.VisualStudio.ErrorHandler;
 
 namespace SonarLint.VisualStudio.Integration.Helpers
 {
     [Export(typeof(IToolWindowService))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
     internal class ToolWindowService : IToolWindowService
     {
-        private readonly IVsUIShell shell;
+        private readonly IVsUIServiceOperation vsUIServiceOperation;
 
         [ImportingConstructor]
-        public ToolWindowService([Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider)
+        public ToolWindowService(IVsUIServiceOperation vsUIServiceOperation)
         {
-            shell = serviceProvider.GetService<SVsUIShell, IVsUIShell>();
-            Debug.Assert(shell != null, "Failed to retrieve the IVsUIShell");
+            this.vsUIServiceOperation = vsUIServiceOperation;
         }
 
         public void Show(Guid toolWindowId)
         {
-            var frame = GetOrCreateWindowFrame(toolWindowId);
-            frame?.Show();
+            vsUIServiceOperation.Execute<SVsUIShell, IVsUIShell>(
+                shell =>
+                {
+                    var frame = GetOrCreateWindowFrame(shell, toolWindowId);
```

Comment:
I looked at refactoring the code to see if there was a way to avoid putting the  `vsUIServiceOperation` call into every interface method implementation, but couldn't see one.